Table block: Fix semantic structure for screen readers on back-end

alexstine opened the pull request:
<!-- Thanks for contributing to Gutenberg! Please follow the Gutenberg Contributing Guidelines:
https://github.com/WordPress/gutenberg/blob/trunk/CONTRIBUTING.md -->

## What?
<!-- In a few words, what is the PR actually doing? -->

Currently, the table block on the back-end is not discoverable as a table by screen readers due to the `role="textbox"` and `contenteditable="true"` attributes set directly on `<td>`. The PR restructures the table so the edit field is contained within the `<td>`.

## Why?
<!-- Why is this PR necessary? What problem is it solving? Reference any existing previous issue(s) or PR(s), but please add a short summary here, too -->

It is important that screen readers behave well with native HTML elements.

## How?
<!-- How is your PR addressing the issue at hand? What are the implementation details? -->

Change HTML structure.

## Testing Instructions
<!-- Please include step by step instructions on how to test this PR. -->
<!-- 1. Open a post or page. -->
<!-- 2. Insert a heading block. -->
<!-- 3. etc. -->

1. Insert a table block and select the Create Table button.
2. Switch into a virtual mode with one of the Windows screen readers. NVDA or JAWS.
3. You should now be able to navigate the table columns and rows with table navigation commands.

### Testing Instructions for Keyboard
<!-- How can you test the changes by using the keyboard only? Please note, this is required for PRs that change the user interface (UI). This ensures the PR can be tested for any possible accessibility regressions. -->

## Screenshots or screencast <!-- if applicable -->

alexstine committed:
Fix table block edit structure for screen readers.

alexstine commented on packages/block-library/src/table/edit.js:
```diff
@@ -414,31 +414,35 @@ function TableEdit( {
-								placeholder={ placeholder[ name ] }
-							/>
+							<CellTag key={ columnIndex }>
+								<RichText
```
This will likely break visuals but it works due to the cell itself remaining a table cell. The editable content is now done within the cell. This will also likely cause a slightly different back-end vs. the front-end but it isn't up for debate. It simply does not work when you convert a table cell to an editable field. The entire table loses semantic meaning with screen readers.

ellatrix replied:
Why do we need role=textbox if it breaks things? Isn't contenteditable enough?

alexstine replied:
@ellatrix With my changes, it comes with 2 bonuses.
1. The table semantics are respected.
2. The editable areas inside the table cells have better text input semantics.

All in all, a big win.

alexstine commented on packages/block-library/src/table/edit.js:
```diff
@@ -414,31 +414,35 @@ function TableEdit( {
 							columnIndex
 						) => (
-							<RichText
-								tagName={ CellTag }
```
No need for this anymore. It renders in `<div>` by default.

alexstine reviewed (commented):
A couple quick implementation notes.

t-hamano commented on packages/block-library/src/table/edit.js:
```diff
@@ -414,31 +414,35 @@ function TableEdit( {
+									scope={
+										CellTag === 'th' ? scope : undefined
+									}
+									colSpan={ colspan }
+									rowSpan={ rowspan }
```
I think the `colSpan` and `rowSpan` props should be moved to the `CellTag` component. Otherwise, I don't think `colSpan` and `rowSpan` will be maintained when the raw data of the table is pasted into the block editor.

alexstine replied:
Done. 👍

t-hamano commented on packages/block-library/src/table/edit.js:
```diff
@@ -414,31 +414,35 @@ function TableEdit( {
-							/>
+							<CellTag key={ columnIndex }>
+								<RichText
+									key={ columnIndex }
```
The key property is required for the `CellTag` component and would not be needed here.

alexstine replied:
Done. 👍

t-hamano commented on packages/block-library/src/table/edit.js:
```diff
@@ -414,31 +414,35 @@ function TableEdit( {
+							<CellTag key={ columnIndex }>
+								<RichText
+									key={ columnIndex }
+									className={ classnames(
+										{
+											[ `has-text-align-${ align }` ]:
+												align,
+										},
+										'wp-block-table__cell-content'
+									) }
```
I think the class name should be given to the `CellTag` component. For example, there may be users who use it as a CSS selector with cell elements, such as `td.has-text-align-center` or `td.wp-block-table__cell-content`.

alexstine replied:
Done. 👍

t-hamano reviewed (commented):
I think this change makes sense. NVDA now reads correctly as follows.

### Before

```
clickable  table  with 0 rows and 0 columns  Body cell text  edit  multi line    ﻿  out of edit  Body cell text  edit  multi line    ﻿  out of edit  Body cell text  edit  multi line    ﻿  out of edit  Body cell text  edit  multi line    ﻿  out of edit  out of table  Table caption text  edit  multi line    ﻿Add caption
```

### After

```
clickable  table  with 2 rows and 2 columns  row 1  column 1  Body cell text  edit  multi line    ﻿  out of edit  column 2  Body cell text  edit  multi line    ﻿  out of edit  row 2  column 1  Body cell text  edit  multi line    ﻿  out of edit  column 2  Body cell text  edit  multi line    ﻿  out of edit  out of table  Table caption text  edit  multi line    ﻿Add caption
```

My guess is that the screen reader is not reading it correctly because there are no cells in the table with the appropriate role attribute.

I also noticed while testing this PR that the first cell is not in focus after inserting the Table block. I think that is one of the reasons why the E2E test is failing. I still don't know why it is not getting focused, but if getting focused makes sense, I think it needs to be resolved.

t-hamano commented:
> I also noticed while testing this PR that the first cell is not in focus after inserting the Table block. I think that is one of the reasons why the E2E test is failing. I still don't know why it is not getting focused, but if getting focused makes sense, I think it needs to be resolved.

To solve this problem, it would be necessary to change the selector of the element that gets focus when the table is created.

https://github.com/WordPress/gutenberg/blob/ab97d28a55b028139b66b8199d795445d830cad8/packages/block-library/src/table/edit.js#L352

As another approach, how about just changing the `role` attribute of the RichText component to `cell`? The text read by the screen reader changes slightly, but the rows and cols appear to be correctly recognized. Is this a problem in terms of accessibility or semantic markup?

alexstine committed:
Fix td props. Fix first cell focus.

alexstine commented:
@t-hamano Thanks for the review.

> As another approach, how about just changing the role attribute of the RichText component to cell? The text read by the screen reader changes slightly, but the rows and cols appear to be correctly recognized. Is this a problem in terms of accessibility or semantic markup?

Not having `role="textbox"` really makes NVDA behave weirdly with `contentEditable`. I would rather do the extra work to keep semantics as close as possible without having to pollute with ARIA.

I fixed the focus code, that was something I implemented. 😄 Really hoping maybe this logic could move into a hook at some point to focus the first element after the block re-renders but don't have the time to scope that out.

alexstine committed:
Move className to td.

t-hamano reviewed (approved):
LGTM 👍

- ✅ After creating the table, the first cell gets focus
- ✅ No visual changes
- ✅ colspan/rowspan is maintained when pasting data containing merged cells from Google Sheets
- ✅ Can still move cells using the arrow keys

ellatrix commented:
What does role=textbox do? Do we need it?

ellatrix commented:
I don't see other editors use this in their editable fields. They just use `contenteditable="true"` and that's it.

alexstine commented:
@ellatrix Here is the doc.

https://developer.mozilla.org/en-US/docs/Web/Accessibility/ARIA/Roles/textbox_role

Other editors should use it if they are not using native inputs. This role adds textbox semantics and `aria-multiline="true"` adds multi-line or text area support. This is all around how screen readers interpret input controls and `contentEditable` only changes the context of the HTML element to be editable vs. the screen reader actually seeing a more accessible/semantic input.

ellatrix commented:
So adding a role makes any element lose its normal semantic meaning? We're doing this for all elements including paragraph, heading etc.

joedolson commented:
> So adding a role makes any element lose its normal semantic meaning? We're doing this for all elements including paragraph, heading etc.

Yes, whenever a role attribute is defined, that changes the semantic definition of the element. The degree to which that matters varies, and a `table` is a case where the semantics are very important in order for the table to be navigable and meaningful for a screen reader. All HTML elements have an implicit role - e.g., the implicit role of a heading is `role="heading"`. Defining an explicit role using the role attribute causes a screen reader to use the replacement semantics instead of the implicit semantics. 

So in order for a screen reader to have an equal experience to the visual parity we're trying to keep for sighted users, the semantics of the elements need to be retained, as well.

alexstine commented:
Let's try this one. Discussion can continue in a follow-up if there are major objections. Thanks! 👍

ellatrix commented:
How does it work properly in TinyMCE without all the role=textbox? https://www.tiny.cloud/docs/tinymce/6/basic-example/

alexstine commented:
@ellatrix It isn't a 1-to-1 comparison. TinyMCE uses a single canvas editing area where all the blocks in Gutenberg are self-contained elements. The best way to compare TinyMCE would be Gutenberg code input mode vs. TinyMCE.

If you inspect the iFrame, the `<body>` tag has `contentEditable="true"` set on it making the whole editor a single pain canvas where this attribute is set per block in Gutenberg. TinyMCE also is not super accessible these days, it has certainly regressed a bit in the way that it is unclear now how to even start editing. The other thing you will notice, since none of the HTML elements have `contentEditable` changing their semantics, each one is read out as an image, heading, table, etc. Voiceover should give you the same experience. I do think that the editor could likely use `aria-multiline="true"` and `role="textbox"` as it would give users a better idea that they are actually in a multi-line text field but that's just me.
